[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #152872
* #152865
* #152730
* #152729
* #152728
* #152727
* #152725
* #152724
* #152704
* __->__ #152626

This PR updates `GuardsStatePickler.reducer_override()` in `torch/_dynamo/guards.py` to handle reconstruction of traceable wrapper subclasses. It's intended to work recursively and handle any level of subclass instance nesting (e.g. subclass instances that contain subclass instances, etc.)

This PR tests the guard on several traceable wrapper tensor subclasses:
* `LocalSubclass`: used to ensure the correct error message is thrown when the subclass is not defined globally
* `torch.testing._internal.two_tensor.TwoTensor`: defines None for its extra metadata
* `SubclassWithMeta`: stores non-trivial extra metadata
* `SubclassWithCustomMetadataGuard`: stores non-trivial extra metadata and defines a custom `__metadata_guard__` classmethod
* `SubclassWithSubclassInnerTensors`: used to test recursiveness; this subclass contains subclass inner tensor components

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jbschlosser committed:
[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH

[ghstack-poisoned]

jbschlosser committed:
Update on "[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zhxchen17 commented on test/dynamo/test_guard_serialization.py:
```diff
@@ -219,6 +383,82 @@ def fn(m, x):
+        from torch.testing._internal.two_tensor import TwoTensor
+
+        tt = TwoTensor(torch.randn(3), torch.randn(3))
+        self._test_serialization("TENSOR_SUBCLASS_METADATA_MATCH", fn, tt)
```
Regarding unittesting, _test_serialization() only tests whether dynamo can finish a roundtrip of the given type of guard. To verify the reconstructed guards have the same behavior, we can also take the returned guard managers from _test_serialization() and use _test_check_fn() to compare the behavior of original guard manager vs deserialized guard manager (like other unittest do). 
Sorry maybe it's not super obvious to do these in two steps, feel free to refactor the unittest code if you have better idea how to test these.

jbschlosser replied:
oh whoops that makes sense. Dumb mistake on my part; will fix. I don't think it's -too- unwieldy to have this in 2 steps, but maybe I'll put a little thought into streamlining this

zhxchen17 commented on torch/_dynamo/guards.py:
```diff
@@ -2524,6 +2541,29 @@ def _unpickle_dispatch_key_set(cls, raw_repr: int):
+                #   (inner attr name, unpickle func, tuple of func inputs)
+                # This supports traceable wrapper subclass inner tensors.
+                inner_data = []
+                attrs, ctx = obj.__tensor_flatten__()
```
Not very familiar with the contract here. I guess we enforce __tensor_flatten__ to be always defined with tensor subclasses? (otherwise we can fire a user friendly error message to remind them to do so)

jbschlosser replied:
Yep that's right - both `__tensor_flatten__()` and `__tensor_unflatten__()` must be defined to be considered a "traceable wrapper subclass". Such tensors must be decomposable into a set of inner tensors.

jbschlosser committed:
Update on "[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH"


This PR updates `GuardsStatePickler.reducer_override()` in `torch/_dynamo/guards.py` to handle reconstruction of traceable wrapper subclasses. It's intended to work recursively and handle any level of subclass instance nesting (e.g. subclass instances that contain subclass instances, etc.)

This PR tests the guard on several traceable wrapper tensor subclasses:
* `LocalSubclass`: used to ensure the correct error message is thrown when the subclass is not defined globally
* `torch.testing._internal.two_tensor.TwoTensor`: defines None for its extra metadata
* `SubclassWithMeta`: stores non-trivial extra metadata
* `SubclassWithCustomMetadataGuard`: stores non-trivial extra metadata and defines a custom `__metadata_guard__` classmethod
* `SubclassWithSubclassInnerTensors`: used to test recursiveness; this subclass contains subclass inner tensor components

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH"


This PR updates `GuardsStatePickler.reducer_override()` in `torch/_dynamo/guards.py` to handle reconstruction of traceable wrapper subclasses. It's intended to work recursively and handle any level of subclass instance nesting (e.g. subclass instances that contain subclass instances, etc.)

This PR tests the guard on several traceable wrapper tensor subclasses:
* `LocalSubclass`: used to ensure the correct error message is thrown when the subclass is not defined globally
* `torch.testing._internal.two_tensor.TwoTensor`: defines None for its extra metadata
* `SubclassWithMeta`: stores non-trivial extra metadata
* `SubclassWithCustomMetadataGuard`: stores non-trivial extra metadata and defines a custom `__metadata_guard__` classmethod
* `SubclassWithSubclassInnerTensors`: used to test recursiveness; this subclass contains subclass inner tensor components

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i

jbschlosser committed:
Update on "[Dynamo] Guard serialization for TENSOR_SUBCLASS_METADATA_MATCH"


This PR updates `GuardsStatePickler.reducer_override()` in `torch/_dynamo/guards.py` to handle reconstruction of traceable wrapper subclasses. It's intended to work recursively and handle any level of subclass instance nesting (e.g. subclass instances that contain subclass instances, etc.)

This PR tests the guard on several traceable wrapper tensor subclasses:
* `LocalSubclass`: used to ensure the correct error message is thrown when the subclass is not defined globally
* `torch.testing._internal.two_tensor.TwoTensor`: defines None for its extra metadata
* `SubclassWithMeta`: stores non-trivial extra metadata
* `SubclassWithCustomMetadataGuard`: stores non-trivial extra metadata and defines a custom `__metadata_guard__` classmethod
* `SubclassWithSubclassInnerTensors`: used to test recursiveness; this subclass contains subclass inner tensor components

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge
